File: src/백도현/week6/6-1/boj2293_dh.java
import java.io.*;
import java.util.*;

public class Main {
    static int N, K;
    static int[] dp, arr;

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        StringTokenizer st = new StringTokenizer(br.readLine(), " ");

		N = Integer.parseInt(st.nextToken());
        K = Integer.parseInt(st.nextToken());
        dp = new int[K+1];
        arr = new int[N];


        for (int i = 0; i < N; i++) {
            arr[i] = Integer.parseInt(br.readLine());
        }
        Arrays.sort(arr);

        dp[0] = 1;
        for (int i = 0; i < N; i++) {
            int x = arr[i];
            for (int j = x; j <= K; j++) {
                dp[j] += dp[j - x];
            }
        }

        bw.write(dp[K] + "\n");

        br.close();
        bw.flush();
        bw.close();
    }

}

File: src/백도현/week6/6-1/boj2293_dh.py
import sys
input = sys.stdin.readline

N, T = map(int, input().split())

arr = []
for _ in range(N):
    arr.append(int(input()))
    
arr.sort()
dp = [0] * (T+1)

dp[0] = 1
for i in range(N):
    for j in range(arr[i], T+1):
            dp[j] += dp[j-arr[i]]
            
print(dp[T])
